code/datums/elements/attachment.dm

```diff
@@ -0,0 +1,36 @@
+/datum/element/attachment
+	element_flags = ELEMENT_BESPOKE
+	id_arg_index = 2
+	///Assoc list of the data required for attaching. It does not change and should not be edited anywhere but here.
+	var/attachment_data
+
+//on_attach, on_detach, on_activate and can_attach are all proc paths that get turned into callbacks when they are called.
+/datum/element/attachment/Attach(datum/target, slot, overlay_icon, on_attach, on_detach, on_activate, can_attach, pixel_shift_x, pixel_shift_y, flags_attach_features, attach_delay, detach_delay, attach_skill, attach_skill_upper_threshold, attach_sound, extra_vars)
+	. = ..()
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+	var/obj/item/target_item = target
+	attachment_data = list(
+		slot = slot, //Slot the attachment fits into, is a string.
+		overlay_icon = overlay_icon ? overlay_icon : target_item.icon, //Icon sheet of the overlay.
+		on_attach = on_attach, //Callback for what the attachment does on attach. Can be null.
+		on_detach = on_detach, //Callback for what the attachment does on detach. Can be null. 
```

Comment:
You regenerate this list every time this element gets attached to a new target
also again will this not have values keyed to what the values are ie

list("string" = "string", /human/type = /human/type) etc